File: 2101-detonate-the-maximum-bombs/2101-detonate-the-maximum-bombs.cpp
using ll = long long;

class Solution {
private:
    int n;
    unordered_map<int, vector<int>> adj;
    unordered_set<int> visited;

public:
    int dfs(int i) {
        visited.insert(i);
        for (int j : adj[i]) {
            if (!visited.count(j)) {
                dfs(j);
            }
        }
        return visited.size();
    }

    int maximumDetonation(vector<vector<int>>& bombs) {
        int n = bombs.size();
        
        for (int i = 0; i < n; ++i) {
            ll xi = bombs[i][0];
            ll yi = bombs[i][1];
            ll ri = bombs[i][2];
            for (int j = 0; j < n; ++j) {
                if (i == j) {
                    continue;
                }
                ll xj = bombs[j][0];
                ll yj = bombs[j][1];
                ll rj = bombs[j][2];
                if (ri*ri >= (xi - xj)*(xi - xj) + (yi - yj)*(yi - yj)) {
                    adj[i].push_back(j);
                    cout << i << ' ' << j << '\n';
                }
            }
        }
        int ans = 0;
        for (int i = 0; i < n; ++i) {
            visited.clear();
            ans = max(ans, dfs(i));
        }
        return ans;
    }
};

File: 2101-detonate-the-maximum-bombs/2101-detonate-the-maximum-bombs.py
class Solution:
    def maximumDetonation(self, bombs: List[List[int]]) -> int:
        n = len(bombs)
        adj = defaultdict(list)
        
        for i in range(n):
            xi, yi, ri = bombs[i]
            for j in range(n):
                if i == j:
                    continue
                xj, yj, rj = bombs[j]
                if ri*ri >= (xi - xj)**2 + (yi - yj)**2:
                    adj[i].append(j)
                    
        def dfs(i):
            visited.add(i)
            for j in adj[i]:
                if j not in visited:
                    dfs(j)
            return len(visited)
            
        ans = 0
            
        for i in range(n):
            visited = set()
            ans = max(ans, dfs(i))
        
        return ans